### PR TITLE
dns_zone: Compute resourceId from the parser, add state migration, switch dns package to generated IDs

### LINKED ISSUE #10778: Error creating azurerm_dns_zone with mixed-case Resource Group: Provider produced inconsistent final plan
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform v0.14.5
+ provider registry.terraform.io/hashicorp/azurerm v2.48.0

### Affected Resource(s)

* `azurerm_dns_zone`

### Terraform Configuration Files

```hcl

provider "azurerm" {
  features {}
  skip_provider_registration = true
}

resource "azurerm_resource_group" "externaldns" {
  name     = "ResGrp_DNS_External"
  location = "Australia East"
}

resource "azurerm_dns_zone" "example-com" {
  name                = "example.com"
  resource_group_name = azurerm_resource_group.externaldns.name
}

resource "azurerm_dns_a_record" "test-example-com" {
  name                = "test"
  zone_name           = azurerm_dns_zone.example-com.name
  resource_group_name = azurerm_dns_zone.example-com.resource_group_name
  ttl                 = 300

  records = ["127.0.0.1"]
}

```

### Expected Behaviour

DNS Zone created for example.com, and A record for test.example.com created.

### Actual Behaviour

```
Error: Provider produced inconsistent final plan

When expanding the plan for azurerm_dns_a_record.test-example-com to include new
values learned so far during apply, provider
"registry.terraform.io/hashicorp/azurerm" produced an invalid new value for
.resource_group_name: was cty.StringVal("ResGrp_DNS_External"), but now
cty.StringVal("resgrp_dns_external").

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
```

### Steps to Reproduce

1. Create ResGrp_DNS_External via Azure Portal
2. `terraform import ResGrp_DNS_External`
3. `terraform apply` to create the DNS Zone and Record

### Important Factoids

Appears to only occur when Resource Group name has mixed or capital case letters.

### References

None.

### PR DESCRIPTION
**Pull request by favoretti**
Fixes #10778

### CODE REVIEW

**Comment by favoretti** [comment]
Complete output is not included, due to DEBUG logging, but complete suite passed.

```
PASS
ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/dns	687.482s
```

File: azurerm/internal/services/dns/dns_a_record_resource.go [hunk 1]
```diff
@@ -133,16 +136,12 @@ func resourceDnsARecordCreateUpdate(d *schema.ResourceData, meta interface{}) er
 		return fmt.Errorf("Error creating/updating DNS A Record %q (Zone %q / Resource Group %q): %s", name, zoneName, resGroup, err)
 	}
 
-	resp, err := client.Get(ctx, resGroup, zoneName, name, dns.A)
+	_, err := client.Get(ctx, resGroup, zoneName, name, dns.A)
 	if err != nil {
 		return fmt.Errorf("Error retrieving DNS A Record %q (Zone %q / Resource Group %q): %s", name, zoneName, resGroup, err)
 	}
```
Comment (tombuildsstuff) [inline azurerm/internal/services/dns/dns_a_record_resource.go:142]: we can remove this Get call altogether now fwiw

File: azurerm/internal/services/dns/dns_aaaa_record_resource.go [hunk 2]
```diff
@@ -137,16 +140,12 @@ func resourceDnsAaaaRecordCreateUpdate(d *schema.ResourceData, meta interface{})
 		return fmt.Errorf("Error creating/updating DNS AAAA Record %q (Zone %q / Resource Group %q): %s", name, zoneName, resGroup, err)
 	}
 
-	resp, err := client.Get(ctx, resGroup, zoneName, name, dns.AAAA)
+	_, err := client.Get(ctx, resGroup, zoneName, name, dns.AAAA)
 	if err != nil {
 		return fmt.Errorf("Error retrieving DNS AAAA Record %q (Zone %q / Resource Group %q): %s", name, zoneName, resGroup, err)
 	}
```
Comment (tombuildsstuff) [inline azurerm/internal/services/dns/dns_aaaa_record_resource.go:146]: (same here) we can remove the redundant Get call here

File: azurerm/internal/services/dns/dns_caa_record_resource.go [hunk 3]
```diff
@@ -138,16 +141,12 @@ func resourceDnsCaaRecordCreateUpdate(d *schema.ResourceData, meta interface{})
 		return fmt.Errorf("Error creating/updating DNS CAA Record %q (Zone %q / Resource Group %q): %s", name, zoneName, resGroup, err)
 	}
 
-	resp, err := client.Get(ctx, resGroup, zoneName, name, dns.CAA)
+	_, err := client.Get(ctx, resGroup, zoneName, name, dns.CAA)
 	if err != nil {
 		return fmt.Errorf("Error retrieving DNS CAA Record %q (Zone %q / Resource Group %q): %s", name, zoneName, resGroup, err)
 	}
```
Comment (tombuildsstuff) [inline azurerm/internal/services/dns/dns_caa_record_resource.go:147]: (same here) we can remove the redundant Get call here

File: azurerm/internal/services/dns/dns_cname_record_resource.go [hunk 4]
```diff
@@ -134,16 +137,12 @@ func resourceDnsCNameRecordCreateUpdate(d *schema.ResourceData, meta interface{}
 		return fmt.Errorf("Error creating/updating CNAME Record %q (DNS Zone %q / Resource Group %q): %s", name, zoneName, resGroup, err)
 	}
 
-	resp, err := client.Get(ctx, resGroup, zoneName, name, dns.CNAME)
+	_, err := client.Get(ctx, resGroup, zoneName, name, dns.CNAME)
 	if err != nil {
 		return fmt.Errorf("Error retrieving CNAME Record %q (DNS Zone %q / Resource Group %q): %s", name, zoneName, resGroup, err)
 	}
```
Comment (tombuildsstuff) [inline azurerm/internal/services/dns/dns_cname_record_resource.go:143]: (same here) we can remove the redundant Get call here

File: azurerm/internal/services/dns/dns_mx_record_resource.go [hunk 5]
```diff
@@ -126,16 +129,12 @@ func resourceDnsMxRecordCreateUpdate(d *schema.ResourceData, meta interface{}) e
 		return fmt.Errorf("Error creating/updating DNS MX Record %q (Zone %q / Resource Group %q): %s", name, zoneName, resGroup, err)
 	}
 
-	resp, err := client.Get(ctx, resGroup, zoneName, name, dns.MX)
+	_, err := client.Get(ctx, resGroup, zoneName, name, dns.MX)
 	if err != nil {
 		return fmt.Errorf("Error retrieving DNS MX Record %q (Zone %q / Resource Group %q): %s", name, zoneName, resGroup, err)
 	}
```
Comment (tombuildsstuff) [inline azurerm/internal/services/dns/dns_mx_record_resource.go:135]: (same here) we can remove the redundant Get call here

File: azurerm/internal/services/dns/dns_ns_record_resource.go [hunk 6]
```diff
@@ -114,16 +117,12 @@ func resourceDnsNsRecordCreate(d *schema.ResourceData, meta interface{}) error {
 		return fmt.Errorf("Error creating DNS NS Record %q (Zone %q / Resource Group %q): %s", name, zoneName, resGroup, err)
 	}
 
-	resp, err := client.Get(ctx, resGroup, zoneName, name, dns.NS)
+	_, err = client.Get(ctx, resGroup, zoneName, name, dns.NS)
 	if err != nil {
 		return fmt.Errorf("Error retrieving DNS NS Record %q (Zone %q / Resource Group %q): %s", name, zoneName, resGroup, err)
 	}
```
Comment (tombuildsstuff) [inline azurerm/internal/services/dns/dns_ns_record_resource.go:123]: (same here) we can remove the redundant Get call here

File: azurerm/internal/services/dns/dns_ptr_record_resource.go [hunk 7]
```diff
@@ -110,16 +113,12 @@ func resourceDnsPtrRecordCreateUpdate(d *schema.ResourceData, meta interface{})
 		return fmt.Errorf("Error creating/updating DNS PTR Record %q (Zone %q / Resource Group %q): %s", name, zoneName, resGroup, err)
 	}
 
-	resp, err := client.Get(ctx, resGroup, zoneName, name, dns.PTR)
+	_, err := client.Get(ctx, resGroup, zoneName, name, dns.PTR)
 	if err != nil {
 		return fmt.Errorf("Error retrieving DNS PTR Record %q (Zone %q / Resource Group %q): %s", name, zoneName, resGroup, err)
 	}
```
Comment (tombuildsstuff) [inline azurerm/internal/services/dns/dns_ptr_record_resource.go:119]: (same here) we can remove the redundant Get call here

File: azurerm/internal/services/dns/dns_srv_record_resource.go [hunk 8]
```diff
@@ -133,16 +136,12 @@ func resourceDnsSrvRecordCreateUpdate(d *schema.ResourceData, meta interface{})
 		return fmt.Errorf("Error creating/updating DNS SRV Record %q (Zone %q / Resource Group %q): %s", name, zoneName, resGroup, err)
 	}
 
-	resp, err := client.Get(ctx, resGroup, zoneName, name, dns.SRV)
+	_, err := client.Get(ctx, resGroup, zoneName, name, dns.SRV)
 	if err != nil {
 		return fmt.Errorf("Error retrieving DNS SRV Record %q (Zone %q / Resource Group %q): %s", name, zoneName, resGroup, err)
 	}
```
Comment (tombuildsstuff) [inline azurerm/internal/services/dns/dns_srv_record_resource.go:142]: (same here) we can remove the redundant Get call here

File: azurerm/internal/services/dns/dns_txt_record_resource.go [hunk 9]
```diff
@@ -119,16 +122,12 @@ func resourceDnsTxtRecordCreateUpdate(d *schema.ResourceData, meta interface{})
 		return fmt.Errorf("Error creating/updating DNS TXT Record %q (Zone %q / Resource Group %q): %s", name, zoneName, resGroup, err)
 	}
 
-	resp, err := client.Get(ctx, resGroup, zoneName, name, dns.TXT)
+	_, err := client.Get(ctx, resGroup, zoneName, name, dns.TXT)
 	if err != nil {
 		return fmt.Errorf("Error retrieving DNS TXT Record %q (Zone %q / Resource Group %q): %s", name, zoneName, resGroup, err)
 	}
```
Comment (tombuildsstuff) [inline azurerm/internal/services/dns/dns_txt_record_resource.go:128]: (same here) we can remove the redundant Get call here

File: azurerm/internal/services/dns/dns_zone_resource.go [hunk 10]
```diff
@@ -177,7 +186,7 @@ func resourceDnsZoneCreateUpdate(d *schema.ResourceData, meta interface{}) error
 		return fmt.Errorf("Error creating/updating DNS Zone %q (Resource Group %q): %s", name, resGroup, err)
 	}
 
-	resp, err := client.Get(ctx, resGroup, name)
+	_, err := client.Get(ctx, resGroup, name)
 	if err != nil {
 		return fmt.Errorf("Error retrieving DNS Zone %q (Resource Group %q): %s", name, resGroup, err)
 	}
```
Comment (tombuildsstuff) [inline azurerm/internal/services/dns/dns_zone_resource.go:192]: (same here) we can remove the redundant Get call here

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @favoretti 

Thanks for pushing those changes - taking a look through on the whole this looks pretty good, but we should be able to remove the redundant `Get` call since we're not using that result as we're creating the ID's here - but if we can fix those up this should otherwise be good to merge 👍 

Thanks!

**Comment by favoretti** [comment]
> hey @favoretti 
> 
> 
> 
> Thanks for pushing those changes - taking a look through on the whole this looks pretty good, but we should be able to remove the redundant `Get` call since we're not using that result as we're creating the ID's here - but if we can fix those up this should otherwise be good to merge 👍 
> 
> 
> 
> Thanks!

Ah, I thought that one served as additional creation verification :)

**Review by tombuildsstuff — APPROVED** [review]
LGTM 👍

**Pull request merged**